Add io.fits test showing that reading from URL works...

drdavella opened the pull request:
This addresses issue #4165 and adds a test to `io.fits` showing that opening a FITS file from a URL already appears to work as expected.

It seems like this one probably doesn't need a change log entry?

EDIT: Fix #4165

pllim commented:
@astrofrog , I set "no change log required" but `astropy-bot` is still not satisfied :question:

bsipocz commented:
@pllim - ``astro-bot`` picked up a new trigger. Probably the milestone change and the new label was added too quickly after each other for it to notice?

bsipocz commented:
Would you be against to backport this to 2.0.x?

pllim commented:
Thanks for the explanation, @bsipocz .

No, I am not against backporting if you don't mind the work. Feel free to re-milestone. :smiley:

bsipocz commented:
@pllim - Doing a clean backport is super easy especially so close to  the first release. If I run into non-trivial issues with the backport, then can always just remilesone it out of the bugfix milestone.

So feel free to mark the doc/test and obviously bug fixes whenever appropriate.

saimn commented:
This seems to work because you are using the `file://` protocol (which is already a good thing), but it fails with a http url:
```python
In [1]: import urllib.request
   ...: from astropy.io import fits

In [2]: obj = urllib.request.urlopen("http://data.muse-vlt.eu/HDFS/v1.0/EmiLines_HDFS_v1p0.fits")

In [3]: fits.open(obj)
---------------------------------------------------------------------------
UnsupportedOperation                      Traceback (most recent call last)
<ipython-input-3-e81a7a49e374> in <module>()
----> 1 fits.open(obj)

~/dev/astropy/astropy/io/fits/hdu/hdulist.py in fitsopen(name, mode, memmap, save_backup, cache, lazy_load_hdus, **kwargs)
    156 
    157     return HDUList.fromfile(name, mode, memmap, save_backup, cache,
--> 158                             lazy_load_hdus, **kwargs)
    159 
    160 

~/dev/astropy/astropy/io/fits/hdu/hdulist.py in fromfile(cls, fileobj, mode, memmap, save_backup, cache, lazy_load_hdus, **kwargs)
    409         return cls._readfrom(fileobj=fileobj, mode=mode, memmap=memmap,
    410                              save_backup=save_backup, cache=cache,
--> 411                              lazy_load_hdus=lazy_load_hdus, **kwargs)
    412 
    413     @classmethod

~/dev/astropy/astropy/io/fits/hdu/hdulist.py in _readfrom(cls, fileobj, data, mode, memmap, save_backup, cache, lazy_load_hdus, **kwargs)
    994             if not isinstance(fileobj, _File):
    995                 # instantiate a FITS file object (ffo)
--> 996                 fileobj = _File(fileobj, mode=mode, memmap=memmap, cache=cache)
    997             # The pyfits mode is determined by the _File initializer if the
    998             # supplied mode was None

~/dev/astropy/astropy/utils/decorators.py in wrapper(*args, **kwargs)
    505                         # one with the name of the new argument to the function
    506                         kwargs[new_name[i]] = value
--> 507             return function(*args, **kwargs)
    508 
    509         return wrapper

~/dev/astropy/astropy/io/fits/file.py in __init__(self, fileobj, mode, memmap, overwrite, cache)
    177             self.size = 0
    178         else:
--> 179             pos = self._file.tell()
    180             self._file.seek(0, 2)
    181             self.size = self._file.tell()

UnsupportedOperation: seek
```

MSeifert04 commented:
@saimn Do you mean that this PR only partially adresses #4165?

drdavella commented:
@saimn, you're right. Is the URL that you provided one that would be reasonable to use in a unit test?

saimn commented:
@MSeifert04 - I checked quickly as I'm trying to catch up with opened issues and PRs, but it looks like the issue is not closed. The current PR should also a test with a remote FITS file.

saimn commented:
@drdavella - If you look at the Astropy tests, there is a mechanism for this kind of tests: with the `@remote_data` decorator, the `get_pkg_data_filename` function, and test files already available on a specific server for Astropy (https://astropy.stsci.edu/data/).

drdavella commented:
It seems like `get_pkg_data_filename` will download the file and provide a link to the cached version, but what I actually need is to use the real URL of the hosted file. Are there any other example tests that do this?

bsipocz commented:
@drdavella - I suggest just use a url directly within a ``remote_data`` decorated test function.

drdavella commented:
Is it necessary to add a specific file to https://astropy.stsci.edu/data for this test, or can I just use one that already exists there?

bsipocz commented:
it's great if you could reuse one from there, but can add another if it's necessary.

drdavella commented:
Okay, I believe this PR now actually addresses the issue as reported. Maybe it should be renamed? Sorry for the earlier run-around. My tests should have been more thorough.

saimn commented on astropy/io/fits/file.py:
```diff
@@ -83,6 +83,7 @@ class _File(object):
     @deprecated_renamed_argument('clobber', 'overwrite', '2.0')
     def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
                  cache=True):
+        from http.client import HTTPResponse
```
While Python 2 compatibility is not dropped, it's better to use `from six.moves import http_client` instead, or we can defer to when Python 2 compatibility is dropped ? Not sure if this PR should or could easily be backported ?

drdavella replied:
Ack, I actually just force-pushed to work around support for Python 2. I could change it to something portable if we wanted to backport, but in my most recent change it just won't work (and won't be tested) for Python 2.

saimn replied:
It's not critical so I think we can skip Python 2 support, but the point is that merging this in the current state may be problematic when backporting other PRs. So maybe maybe we just wait a bit before merging ?

drdavella replied:
It's not a big deal at all for me to do the portable thing, so I can just update it to use `six.moves` for now.

saimn commented on astropy/io/fits/tests/test_core.py:
```diff
@@ -593,6 +595,36 @@ def test_open_nonexistent(self):
+    @remote_data(source='astropy')
+    def test_open_from_remote_url(self):
+        import urllib.request
+        remote_url = 'https://astropy.stsci.edu/data/allsky/allsky_rosat.fits'
```
Could you use the `dataurl` config item instead ? To avoid having a hard-coded value, which may change in the future.

saimn replied:
Also the two tests can probably be merged to avoid repetition.

drdavella replied:
I will push an update using `dataurl` to get the base URL path. I might argue in favor of keeping the tests separate since they test fairly different code paths (as you discovered when you tried to test a remote URL). However I'll defer to your judgment.

bsipocz commented:
@astrofrog - the bot should have complained here, as there is a changelog but also the ``no-changelog...`` label. I'm removing the label, as indeed a changelog entry won't hurt here as this PR fixes an existing issue.

MSeifert04 commented on astropy/io/fits/file.py:
```diff
@@ -119,9 +120,16 @@ def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
+            mode not in ('ostream', 'append', 'update') and _is_url(fileobj)):
+            self.name = download_file(fileobj, cache=cache)
+        # Handle responses from URL requests that have already been opened
+        elif isinstance(fileobj, http_client.HTTPResponse):
```
Why not just check for an `url` attribute? That would also allow `requests` objects. :)

Also it's weird to extract the url from an already open ressource and make a new request instead of using the data provided by the url-object

MSeifert04 replied:
I mean someting like `fits.HDUList.fromstring(fileobj.content)` (`requests`) or `fits.HDUList.fromstring(obj.read())` (`urlopen` objects).

drdavella replied:
I definitely agree with what you're saying about opening a new resource based on an already open one. It seemed to require the least amount of new code, but there's probably a better way.

drdavella replied:
Also it does seem like a good idea handle `Request` as well, but I don't think it's as simple as checking for `hasattr(fileobj, 'url')` since I think the `Request` attribute is `full_url`. It will still be necessary to test for `isinstance(..., HTTPResponse)` and `isinstance(..., Request)` since they will need to be handled differently.

drdavella replied:
I added support for `Request` objects. After a little more investigation, I think that even though calling `download_file` opens another resource, it's still a fairly clean implementation. Any alternative would have to store the contents of the `HTTPRequest` to a temporary file and open that, so it would still involve opening and closing other resources. Or it would require a fairly substantial refactor of `download_file`, which might actually be the cleanest solution.

bsipocz commented on CHANGES.rst:
```diff
@@ -172,6 +172,9 @@ astropy.io.ascii
 ^^^^^^^^^^^^^^^
 
+- Properly handle opening of FITS files from HTTPResponses from remote URLs.
+  [#6398]
```
Please fix the typo here in the PR number. Also while you're at it, move the entry down to the 2.0.1 section.

drdavella replied:
Done, sorry bout that.

bsipocz commented on astropy/io/fits/tests/test_core.py:
```diff
@@ -593,6 +595,48 @@ def test_open_nonexistent(self):
             assert os.path.exists(self.temp('foobar.fits'))
             os.remove(self.temp('foobar.fits'))
 
+    @pytest.mark.skipif(six.PY2,
```
this is not required any more, isnt it? since you now use urllib from six.moves

drdavella replied:
Oddly enough, it looks like you still can't use `urllib` from `six.moves` exactly the same way in Python 2.7 and 3. For example, in Python 3 I can use the result of `urllib.request.urlopen` in a `with` context, but I can't do that in Python 2.7. I can work around this with some version-specific code in the unit tests if you want, but I'm not sure it's worth it.

bsipocz replied:
Thanks for the explanation. This case I think it's fine to skip the tests.

saimn replied:
Ok for skipping the test, but in this case it would be hazardous to backport this change, so maybe we just re-milestone this to 3.0 ?

drdavella replied:
I'll leave this decision up to someone else but I'd be happy to add version-specific logic if it makes it easier to integrate the changes sooner.

drdavella commented:
Sorry this PR has turned into a bit of a mess. I can try to squash and clean up once the tests are all passing and I have addressed all of the code review comments.

I'm trying to track down the current test failure, which looks like this:
> ________________________ test_skip_remote_data_astropy _________________________
> [snip]
> IOError: An attempt was made to connect to the internet by a test that was not marked `remote_data`. The requested host was: astropy.org

I was able to reproduce this error locally by running the following command in a Python 2.7 environment:
`$ ./setup.py test --open-files --remote-data=astropy -t astropy/tests/tests/test_skip_remote_data.py`

However, it occurs only sporadically, and after running the command a bunch of times I actually can't reproduce it at all anymore.

I sort of suspect that this error has nothing to do with the changes that I just made, but I can't confirm since it's hard to reproduce. I also suspect that if I push more commits to address code review comments, we might not see this failure again.

Does this look familiar to anyone else? If not, I can open a separate issue.

saimn commented:
The test failure seems unrelated, and it passed with your latest commits.

saimn commented:
After thinking a bit more about this, and experimenting a bit, I'm not sure the currently proposed solution is the best way to go:

For `urllib.request.Request`, i.e. requests that are not yet done: the solution is fine but the interest is small. Constructing "manually" a `Request` is probably interesting only to make special things like handling authentication. But in this case the solution of getting the url may not work. 

So I think the more interesting usecases would be to use `urlopen`, to make the request, either with a string url, or with a special Request (handling authentication, cookies, whatever), and provide the result of this to Astropy: a `http.client.HTTPResponse`.

Then for `http.client.HTTPResponse` objects, the request is already done, there is no need to do it twice (and it may fail for authenticated requests). This is where we could something more interesting, using directly the result of this request, by fixing the `UnsupportedOperation: seek` error shown [above](https://github.com/astropy/astropy/pull/6378#issuecomment-317458154). The error happens because `http.client.HTTPResponse` subclass `io.BufferedIOBase` which does not implement `seek` and `tell`. I see two options:

- wrap the response result with `io.BytesIO`:
```
In [62]: o = urllib.request.urlopen('http://astropy.stsci.edu/data/photometry/irac_ch1_flight.f
    ...: its')

In [65]: oo = io.BytesIO(o.read())

In [66]: fits.open(oo)
Out[66]: [<astropy.io.fits.hdu.image.PrimaryHDU object at 0x7f592af3a400>]
```
- or use `HDUList.fromstring` with the response bytes:
```
In [67]: o = urllib.request.urlopen('http://astropy.stsci.edu/data/photometry/irac_ch1_flight.f
    ...: its')

In [68]: fits.HDUList.fromstring(o.read())
Out[68]: [<astropy.io.fits.hdu.image.PrimaryHDU object at 0x7f592ac4dd30>]
```

Not sure which is best, the first one can be implemented in `_File`, which allows to keep all file reading stuff in the same place, and the second one looks cleaner but must be done in `fitsopen`, to use `HDUList.fromstring` instead of `HDUList.fromfile`.

drdavella commented:
@saimn I think the method using `BytesIO` makes good sense and I updated the response handling logic to use that instead of `download_files` which, as you pointed out, is problematic. I left the request handling logic in place, although I agree that it seems to be of limited value.

In addition, I squashed everything a bit to make things look a little cleaner.

saimn commented on astropy/io/fits/file.py:
```diff
@@ -119,10 +120,24 @@ def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
+            if mode in ('ostream', 'append', 'update'):
+                raise ValueError(
+                    "Mode {} not supported for HTTPResponse".format(mode))
+            from io import BytesIO
```
*nitpick*: move the import at the top

drdavella replied:
The reason It's not at the top is because you can avoid the import in the failure case. I'm not sure it matters much either way.

saimn replied:
It's from the stdlib and surely already imported (it's the module that provides `open` ;-)).

drdavella replied:
Sorry, do you mean to the top of that `elif` or to the top of the file?

saimn replied:
The top of the file, all imports should be global unless there is a good reason (circular imports, optional dependencies).

saimn commented on CHANGES.rst:
```diff
@@ -253,6 +253,9 @@ astropy.io.ascii
 ^^^^^^^^^^^^^^^
 
+- Properly handle opening of FITS files from URL Requests and HTTPResponses
+  from remote URLs. [#6378]
```
*Requests* and *HTTPResponses* with the plural form may be confusing. If we care only about Python 3, you could use instead the complete name (`http.client.HTTPResponse`) or just mention it as the *output of `urlopen`* ?

saimn commented:
Thanks @drdavella , I changed the milestone to 3.0 hence we don't need to bother too much with Python 2 (we just need to keep the code compatible until Python 2 compatibility is removed). Can you move the changelog entry to 3.0 ? For Request objects, I would vote for removing this as I don't see how it can be useful, but it's up to you ;-)

drdavella commented:
@saimn okay I think I've addressed all the feedback:
- fixed import statement
- clarified change log comment and moved to 3.0.0
- removed support (and test) for `HTTPRequest`

saimn commented on astropy/io/fits/file.py:
```diff
@@ -20,7 +21,8 @@
                    _array_to_file, _write_string)
-from ...extern.six import b, string_types
+from ...extern.six.moves import http_client
+from ...extern.six import b, string_types, moves
```
`moves` import is not used.

saimn commented:
ping @MSeifert04, if you want to have another look ?

MSeifert04 commented on astropy/io/fits/file.py:
```diff
@@ -119,10 +121,16 @@ def __init__(self, fileobj=None, mode=None, memmap=None, overwrite=False,
+            if mode in ('ostream', 'append', 'update'):
+                raise ValueError(
+                    "Mode {} not supported for HTTPResponse".format(mode))
+            fileobj = io.BytesIO(fileobj.read())
```
Not really a good idea because once `read` subsequent `read`s will return an empty string. I wouldn't expect `open` to "clear" the Response object.

saimn replied:
If you pass a Response object, I think you have to expect that it will be used to read the data. Otherwise there are other ways, passing the url or saving the data to disk.

drdavella replied:
FWIW the original solution using `download_files` actually avoided this problem, even if it was at the cost of redundant resources being used. However, I think I agree with @saimn that the current behavior is reasonable given the use case.

MSeifert04 commented:
I'm not sure if we actually should support open "response" objects. We already support reading from urls. Wouldn't it make sense to just error out in case a response is passed in with a helpful message (pass in the url instead)?

saimn commented:
The question is to know if there is a real usecase for this. For HTTP responses, it is logical to support it because it is a file-like object, which is already supported by `fits.open`. And with this change, if you really want to use a Request object, you can use `fits.open(urlopen(request))`.

drdavella commented:
@saimn and @MSeifert04, it's easy enough for me to add back if you want. On the one hand, I agree with @saimn that there's probably not a real use case for this. On the other hand, I agree with @MSeifert04 that it's not that hard to support requests and it makes things seem more 'complete'. At this point I don't have strong feelings about this so I'll just wait for someone else to tell me what to do :sweat_smile:.

MSeifert04 commented:
@saimn But https://github.com/astropy/astropy/pull/6378#issuecomment-317458154 shows that while it's somewhat file-like it's not exactly file-like.

I'm not sure either. It probably won't hurt to support requests but I'm a bit hesitant because it seems like we're trying to fix a non-existing problem.

saimn commented:
@MSeifert04 :

> But #6378 (comment) shows that while it's somewhat file-like it's not exactly file-like.

Yes, for `HTTPResponse`, hence the fix implemented currently. It allows to support the output of `urlopen`.

> It probably won't hurt to support HTTP responses but I'm a bit hesitant because it seems like we're trying to fix a non-existing problem.

I'm confused by this, do you mean for responses, requests, or both ? For responses, as show above, it's almost file-like and I think it makes sense to support this. But for requests I'm also advocating that it's a non-existing problem.

MSeifert04 commented:
I'm not sure what you mean by "requests". Personally I feel that passing an url to `fits.open` is a "request": `fits.open("http://data.muse-vlt.eu/HDFS/v1.0/EmiLines_HDFS_v1p0.fits")`. Or did you mean the `requests` python package?

As I see it there is only one use-case to pass in responses: When the online material requires authentication, because that way you can't just pass in the url to `fits.open`. But for authentication it  would be much safer and better to use the `requests` package or `astroquery` (if it's supported there).

I'm not totally against supporting `urlopen` but I think it should rather error out (with a message that one should use `HDUList.fromstring(whatever.read())` than to implement it ourselves.

saimn commented:
@MSeifert04 - i mean `urllib.request.Request`, which is something you can  pass to `urllib.request.urlopen`. But actually I read *Request* instead of *Response* in your [comment](https://github.com/astropy/astropy/pull/6378#issuecomment-318556845) above, I think this explain the confusion. 

To clarify:

- the current PR allows to pass a `http.client.HTTPResponse` to `fits.open`. This is returned by `urlopen`, and is almost file-like (it misses the `seek` and `tell` methods). As `fits.open` supports file-like objects, and as using `HDUList.fromstring(whatever.read())` is not easy to discover, I think it makes sense to support this.

- at some point @drdavella also added the possibility to pass a `urllib.request.Request` to `fits.open`, and I thought you were speaking about this.

drdavella commented:
@MSeifert04 
> I'm not totally against supporting urlopen but I think it should rather error out (with a message that one should use HDUList.fromstring(whatever.read()) than to implement it ourselves.

Couldn't this argument also apply to results from `open`, which is already supported? The cases seem to be analogous. Personally I think it's reasonable to add support for `urlopen`, especially since this behavior was explicitly requested in the issue this PR attempts to resolve.

MSeifert04 commented:
> especially since this behavior was explicitly requested in the issue this PR attempts to resolve.

Ah, I thought that issue was just about the "misleading" error. Not about supporting it (although it was mentioned that it would be easy).

But I don't feel strong about that so if you actually want to support it I won't mind. 😄

drdavella commented:
Again, I realize I'm a biased source, but it's easy to support and it's consistent with the way that regular file objects are already supported, so I think it makes sense to add support for objects from `urlopen`. It also seems like it's hopefully useful to someone out there (based on the SO question). It's already implemented now, and it works, so I say leave it.

But I'm new here, so I'll leave the final decision to someone else :zipper_mouth_face:

saimn commented:
@MSeifert04 - With the full quote it suggests that it could be supported:

>  A urlopen object has never been supported (intentionally) in the first place. But it could be supported easily so long as the file is opened in readonly (or denywrite) mode. It just needs to be handled explicitly. In any case the error message one gets is also misleading:

Also the issue was triggered by https://stackoverflow.com/questions/32643206/how-do-i-open-a-fits-file-from-a-url-in-astropy where someone tries to do exactly this. 

So, I think everybody agree that we spent already too much time discussing this :wink:. It's not a major feature but can be useful for specific cases. Let's merge and move forward (is it a special day where everybody reports FITS issues ? :open_mouth: ).

saimn commented:
And thanks @drdavella !
